File: LightController/pca9685.js


module.exports = function(address){
    var rasp2c = null;
    try {
        rasp2c = require('rasp2c');
    }
    catch (ex) {
        console.log('rasp2c is not available on this system.  Using emulation mode instead.');
        var channels = [0, 0, 0, 0, 0, 0];
        return {

            setChannel: function (channel, value) {
                console.log('Setting ' + channel + ' to ' + value);
                channels[channel] = value;
            },
            reset: function () {
                console.log('reset');
            },
            allOff: function () {
                console.log('all off');
                for(var i = 0; i < channels.length; i++)
                    channels[i] = 0;
            },
            allOn: function () {
                console.log('all on');
                for (var i = 0; i < channels.length; i++)
                    channels[i] = 1;
            },
            setPWMFreq: function (freq) {
                console.log('Setting frequency to ' + freq);
            },
            setPWM: function (pin, on, off) {
                console.log('Setting pwm on pin ' + pin + '. On: ' + on + 'Off: ' + off);
                channels[pin] = ((off - on) / 4095);
            },
            setChannel: function (pin, val) {
                console.log('Setting channel ' + pin + ' to ' + val);
                channels[pin] = val;
            },
            channelOn: function (pin) {
                console.log('Channel ' + pin + ' turned on');
                channels[pin] = 1;
            },
            channelOff: function (pin) {
                console.log('Channel ' + pin + ' turned off');
                channels[pin] = 0;
            },
            readPin: function (pin, cb) {
                console.log('Reading channel ' + pin + '.');
                if (cb) {
                    cb(channels[pin]);
                }
                else {
                    console.log(channels[pin]);
                }
            },
            fadeTo: function (pin, val) {
                console.log('Fading channel ' + pin + ' to ' + val);
            }
        };
    }

	var addr = 0x64;
	if (address){
		addr = address;
	}
    
    var PCA9685_MODE1 = 0x0;
	var PCA9685_PRESCALE = 0xFE;
    
    var BASE_LED_ADDR = 0x06;
	var ALL_LED_ADDR = 0xFA;
	
	function write_cb(err, result){
		if (err){
			console.log('ERROR: ');
			console.log(err);
		}
	}
	
	function fadeTo(channel, val){
		readPin(channel, function(start){
			fadeChannel(channel, start, val, .001, 500);
		});
	}
	
	function fadeChannel(channel, start, end, step, speed){
		if (start < end){
			start+=step;
			setPin(channel, start);
			if (start < end){
				setTimeout(function(){fadeChannel(channel, start, end, step, speed);}, speed);
			}
		}
		else{
			start-=step;
			setPin(channel, start);
			if (start > end){
				setTimeout(function(){fadeChannel(channel, start, end, step, speed);}, speed);
			}			
		}
	}
	
	function reset(){
		rasp2c.set(addr, PCA9685_MODE1, 0x0, write_cb);
	}

	function setPWMFreq(freq) {
		freq *= 0.9;
		var prescale = 25000000;
		prescale /= 4096;
		prescale /= freq;
		prescale -= 1;
		  
		prescale = Math.floor(prescale + 0.5);
		rasp2c.dump(addr, PCA9685_MODE1 + '-' + PCA9685_MODE1, function(err, result){
			if (err){
				
			}
			else{
				var oldmode = parseInt(result);
				var newmode = (oldmode&0x7F) | 0x10; // sleep
				rasp2c.set(addr, PCA9685_MODE1, newmode, write_cb); // go to sleep
				rasp2c.set(addr, PCA9685_PRESCALE, prescale, write_cb); // set the prescaler
				rasp2c.set(addr, PCA9685_MODE1, oldmode, write_cb);
				setTimeout(function(){
					rasp2c.set(addr, PCA9685_MODE1, oldmode | 0xa1, write_cb);  
				}, 50);
			}
		});								  
    }
    
    function setAll(on, off){
        var off = 4096 * val;
        var data_addr = ALL_LED_ADDR;
        rasp2c.set(addr, data_addr++, 0, write_cb);
        rasp2c.set(addr, data_addr++, 0, write_cb);
        rasp2c.set(addr, data_addr++, off & 0xFF, write_cb);
        rasp2c.set(addr, data_addr++, off >> 8, write_cb);
    }
	
    function allOn(){
        setAll(0, 1);
	}
	
    function allOff(){
        setAll(0, 0);
	}
	
	function pinOn(pin){
		setPWM(pin, 4096, 0);
	}
	
	function pinOff(pin){
		setPWM(pin, 0, 4096);
	}
	
	function readPin(pin, callback){
		var startAddr = (pin*4) + BASE_LED_ADDR;
		var endAddr = startAddr + 3;
		rasp2c.dump(addr, startAddr + '-' + endAddr, function(err, val){
			if (err){
				console.log('Error: ' + err);
			}
			else{
				var on = (val[1] << 8) + val[0];
				var off = (val[3] << 8) + val[2];
                
                var pct = 0;
                if (on == 4096 && off == 0) {
                    pct = 1;
                }
                else if (on == 0 && off == 4096) {
                    pct = 0;
                }
                else {
                    pct = (off - on) / 4096;
                }
                
                if (callback) {
                    callback(pct);
                }
                else {
                    console.log(pct);
                }

			}
		});
	}
	
	function setPin(pin, val){
		if ((pin < 0) || (pin > 15)){
			throw new Error('pin must be between 0 and 15.');
		}
		if ((val < 0) || (val > 100)){
			throw new Error('val should be a percentage between 0 and 100.');
		}
		if (val > 1){
			val/=100;
		}
		setPWM(pin, 0, Math.round(4095 * val));
	}
	
	function setPWM(pin, on, off){
		var data_addr = (pin*4) + BASE_LED_ADDR;
		rasp2c.set(addr, data_addr++, on & 0xFF, write_cb);
		rasp2c.set(addr, data_addr++, on >> 8, write_cb);
		rasp2c.set(addr, data_addr++, off & 0xFF, write_cb);
		rasp2c.set(addr, data_addr++, off >> 8, write_cb);
	}
	
	return {
		reset: reset,
		setPWMFreq: setPWMFreq,
		setPWM: setPWM,
		setChannel: setPin,
		channelOn: pinOn,
		channelOff: pinOff,
		allOff: allOff,
		allOn: allOn,
		fadeChannel: fadeChannel,
		readPin: readPin,
		fadeTo: fadeTo
	};
	
}

File: RPIPWMDriverFirmware/RPIPWMDriverFirmware/ArduinoCore/src/core/PCA9685.cpp
/*
 * PCA9685.cpp
 *
 * Created: 3/27/2016 3:55:44 PM
 *  Author: Michael
 */ 


#include <PCA9685.h>
#include <Wire.h>

PCA9685::PCA9685(uint8_t address) {
  _address = address;
}

void PCA9685::begin(void) {
 Wire.begin();
 delay(10);
 setAll(0);
 reset();
}


void PCA9685::reset(void) {
 write8(PCA9685_MODE1, 0x20);
   Wire.beginTransmission(_address);
   Wire.write(PCA9685_MODE1);
   Wire.write(0x20);				//Auto Increment
   Wire.endTransmission();
}

void PCA9685::setFrequency(float frequency) {
  frequency *= 0.9;  // Correct for overshoot in the frequency setting (see issue #11).
  float prescaleval = 25000000;
  prescaleval /= 4096;
  prescaleval /= frequency;
  prescaleval -= 1;
  uint8_t prescale = floor(prescaleval + 0.5);
  
  uint8_t oldmode = read8(PCA9685_MODE1);
  uint8_t newmode = (oldmode&0x7F) | 0x10; // sleep
  write8(PCA9685_MODE1, newmode); // go to sleep
  write8(PCA9685_PRESCALE, prescale); // set the prescaler
  write8(PCA9685_MODE1, oldmode);
  delay(5);
  write8(PCA9685_MODE1, oldmode | 0xa1);  //  This sets the MODE1 register to turn on auto increment.
                                          // This is why the beginTransmission below was not working.
}

void PCA9685::allOn()
{
	setAll(4095);	
}

void PCA9685::allOff()
{
	setAll(0);
}

void PCA9685::on(uint8_t channel)
{
	setPin(channel, 4095);
}

void PCA9685::off(uint8_t channel)
{
	setPin(channel, 0);
}

void PCA9685::setAll(long value)
{
		value = min(value, 4095);
		setPin(61, value);
}

void PCA9685::setChannel(uint8_t channel, long value)
{
	setPin(channel, value);
}

void PCA9685::setChannels(long* values)
{
	uint8_t buff[16*4];
	uint16_t val;
	
	for(uint8_t i = 0; i < NUM_CHANNELS; i++)
	{
		values[i] = min(values[i], 4095);
		
		if (values[i] == 4095) {
			// Special value for signal fully on.
			buff[(i*4)] = 4096;
			buff[(i*4)+1] = 4096>>8;
			buff[(i*4)+2] = 0; 
			buff[(i*4)+3] = 0;
		}
		else if (values[i] == 0) {
			// Special value for signal fully off.
			buff[(i*4)] = 0;
			buff[(i*4)+1] = 0;
			buff[(i*4)+2] = 4096; 
			buff[(i*4)+3] = 4096>>8;
		}
		else{
			buff[(i*4)] = 0;
			buff[(i*4)+1] = 0;
			buff[(i*4)+2] = values[i]; //Will this be a bug?
			buff[(i*4)+3] = values[i]>>8;
		}
	}
	
	Wire.beginTransmission(_address);
	Wire.write(LED0_ON_L);
	for(uint8_t i = 0; i < NUM_CHANNELS*4; i++)
	{
		Wire.write(buff[i]);
	}
	Wire.endTransmission();
	
}

long PCA9685::readChannel(uint8_t pin)
{
	if (pin >= NUM_CHANNELS)
		return 0;
		
	uint8_t startAddr = (pin*4) + BASE_LED_ADDR;
	
	long on;
	long off;
	byte val[4];
	uint8_t i = 0;
	
	Wire.beginTransmission(_address);
	Wire.write(startAddr);
	Wire.endTransmission();

	Wire.requestFrom(_address, 4);
	while (Wire.available())
	{
		val[i++] = Wire.read();
	}
	on = (val[1] << 8) + val[0];
	off = (val[3] << 8) + val[2];
	if (off == 4096)
		return 0;
	if (on == 4096)
		return 4095;
	
	return off;
	
}

byte PCA9685::readRegister(uint8_t location)
{
	byte val;
	Wire.beginTransmission(_address);
	Wire.write(location);
	Wire.endTransmission();

	Wire.requestFrom(_address, 1);
	while (Wire.available())
	{
		val = Wire.read();
	}
	return val;
}

void PCA9685::setPWM(uint8_t num, uint16_t on, uint16_t off) {
	
	Wire.beginTransmission(_address);
	Wire.write(LED0_ON_L+4*num);
	Wire.write(on);
	Wire.write(on>>8);
	Wire.write(off);
	Wire.write(off>>8);
	Wire.endTransmission();
}

// Sets pin without having to deal with on/off tick placement and properly handles
// a zero value as completely off.  Optional invert parameter supports inverting
// the pulse for sinking to ground.  Val should be a value from 0 to 4095 inclusive.
void PCA9685::setPin(uint8_t num, uint16_t val)
{
	// Clamp value between 0 and 4095 inclusive.
	val = min(val, 4095);
	
	if (val == 4095) {
		// Special value for signal fully on.
		setPWM(num, 4096, 0);
	}
	else if (val == 0) {
		// Special value for signal fully off.
		setPWM(num, 0, 4096);
	}
	else {
		setPWM(num, 0, val);
	}
	
}

uint8_t PCA9685::read8(uint8_t addr) {
  Wire.beginTransmission(_address);
  Wire.write(addr);
  Wire.endTransmission();

  Wire.requestFrom((uint8_t)_address, (uint8_t)1);
  return Wire.read();
}

void PCA9685::write8(uint8_t addr, uint8_t d) {
  Wire.beginTransmission(_address);
  Wire.write(addr);
  Wire.write(d);
  Wire.endTransmission();
}
